### PR TITLE
services.xserver: assert that either desktop- or window manager is not "none"

### LINKED ISSUE #11064: Assert when no window manager nor desktop manager are enabled.
The goal is to ensure that NixOS users do not forget to enable a window manager or a desktop manager when they enable the X server.  As soon as we have a didactic message, we can change the default value of `services.xserver.desktopManager.default` to replace it by `"none"` and change the example value to be more talkative.

To do that, we should add an `assertions` or a `warning` into display-managers/default.nix module [1] which checks the sessions options of the desktop manager [2] and the session option of the window manager [3] and report if these are empty.

[1] https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix 
[2] https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/desktop-managers/default.nix#L29
[3] https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/window-managers/default.nix#L36

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

setting a DM or WM is something I tend to forget when I want to built test machines with an X11 inside, so this assertion might catch unexpected behavior when building machines with X.

resolves #11064 (/cc @nbp )

I tested this against the following expressions:

``` nix
{
  # VMs without X11 still evaluate
  empty = { pkgs, ... }: {
  };

  # needs to be commented out to avoid evaluation errors
  # no DM/WM set, should break
  onlyX11 = { pkgs, ... }: {
    services.xserver.enable = true;
  };

  # DM set, evaluates
  x11WithDM = { pkgs, ... }: {
    services.xserver = {
      enable = true;
      desktopManager.plasma5.enable = true;
      desktopManager.default = "plasma5";
    };
  };

  # WM set, evaluates
  x11WithWM = { pkgs, ... }: {
    services.xserver = {
      enable = true;
      windowManager.i3.enable = true;
      windowManager.default = "i3";
    };
  };

  # plain X sessions must be enabled explicitly
  x11Plain = { pkgs, ... }:{
    services.xserver.enable = true;
    services.xserver.plainX = true;
  };
}
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by rycee** [comment]
I'm not fond of this. Especially since it breaks my setup, which uses `~/.xsession`. My complete `services.xserver` configuration for my laptop, including irrelevant stuff is
```
  services.xserver = {
    enable = true;
    layout = "us";
    displayManager.lightdm.enable = true;
    monitorSection = ''
      # Set the right DPI. Screen is 344x193 mm.
      DisplaySize 344 193
    '';
    libinput.enable = true;
  };
```

**Comment by vyp** [comment]
But what if we just want to use a plain X session?

**Comment by Ma27** [comment]
interesting, didn't know there are people using X without a DM or a WM. Just out of interest: what's the reason for a plain X session?

**Comment by rycee** [comment]
I'm using Home Manager to manage my X session. It mainly uses systemd user services started from [~/.xsession](https://github.com/rycee/home-manager/blob/2ff8c12bf96b40c2639057f2a89ec12aad369fe0/modules/xsession.nix#L142-L174).

**Comment by vyp** [comment]
@Ma27 I'm not using without a DM/WM, but I find it quite useful as a debug step sometimes to figure out if an issue is due to the window manager, or say a regression in X sometimes (and obviously knowing that can save a lot of time). Also helps with debugging say a video driver with not so good linux kernel support, since you can rule out potential effects of DM/WM. (And the NixOS generations feature makes this easy to test as well.) I'm not opposed to say some configuration option that allows this, since my use case is admittedly rare, I just don't want it to be impossible.

**Comment by Ma27** [comment]
@vyp @rycee thanks a lot for your explanations.

> I'm not opposed to say some configuration option that allows this

so you mean a configuration option like `services.xserver.plain` which defaults to `false` and needs to be set to `true` explicitly?

**Comment by vyp** [comment]
> so you mean a configuration option like `services.xserver.plain` which defaults to `false` and needs to be set to `true` explicitly?

Yep :+1:

**Comment by Ma27** [comment]
@rycee if you're fine with this, I'd implement that additional option.

If not, I'd consider closing this for now :-)

**Comment by Ma27** [comment]
@vyp  just found some time to implement this. Would you mind having a look at the change to confirm that this is a better solution? :)

**Comment by rycee** [comment]
Yeah, I think this is a decent compromise. I would prefer a name more along the lines of `allowXSessionFile` or `allowXSessionOverride` but it doesn't really matter.

**Comment by vyp** [comment]
@Ma27 unfortunately am really busy (time-sensitive) for at least a few days so afraid not right now. :/ (only just have time for making short responses from my phone when traveling tbh, if you're curious)

**Comment by Ma27** [comment]
rebased onto latest master which should fix the build.

> Yeah, I think this is a decent compromise. I would prefer a name more along the lines of allowXSessionFile or allowXSessionOverride but it doesn't really matter.

I don't like this as setting a DM/WM doesn't prohibit overriding the XSession file. The only reason why this option exists is to "tell" Nix whether plain X sessions without a DM/WM are allowed.

>  (only just have time for making short responses from my phone when traveling tbh, if you're curious)

ok, I'll try to summarize it for you: I added an option `services.xserver.plainX` (default=false) which allows to set neither a WM nor a DM if true.

**Comment by rycee** [comment]
I'm fine with this. I am thinking that it warrants a release note since it is a change that will break some people's installs. In particular, it should be clear what steps you need to take (enable `services.xserver.plainX`) to get things working again.

**Comment by Ma27** [comment]
makes sense, will add this :-)

**Comment by Ma27** [comment]
@rycee done :-)

**Comment by rycee** [comment]
Thanks!

File: nixos/modules/services/x11/xserver.nix [hunk 1]
```diff
@@ -161,6 +161,14 @@ in
         '';
       };
 
+      plainX = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether the X11 session can be plain (without DM/WM) or not.
```
Comment (nbp) [inline nixos/modules/services/x11/xserver.nix]: This description should explain that the Xsession is then used as a fallback.

**Review by nbp — APPROVED** [review]

**Comment by Ma27** [comment]
@nbp done :-)

File: nixos/modules/services/x11/xserver.nix [hunk 2]
```diff
@@ -552,6 +561,9 @@ in
                 + "${toString (length primaryHeads)} heads set to primary: "
                 + concatMapStringsSep ", " (x: x.output) primaryHeads;
       })
+      { assertion = cfg.desktopManager.default == "none" && cfg.windowManager.default == "none" -> cfg.plainX;
+        message = "Either the desktop manager or the window manager shouldn't be `none`!";
```
Comment (nbp) [inline nixos/modules/services/x11/xserver.nix]: The best error message is the error message which let you fix the issue which is being reported.
This message should include some hint to fix the configuration.

Comment (Ma27) [reply nixos/modules/services/x11/xserver.nix]: agreed. Added another sentence which mentions the `plainX` option ...

**Review by nbp — APPROVED** [review]

**Comment by NeQuissimus** [comment]
hmm, I am quite unhappy with this (since my rebuild just broke)
If I enable only a single windowmanager (i.e. xmonad), why would I need to explicitly define `default = "xmonad"`?

Some sort of default should be figured out automatically. 
Can we give the modules something like `default = mkDefault "xmonad";` ?

**Comment by Ma27** [comment]
I'm not 100% sure what you mean, sorry.
Could you please share the relevant parts of you config? :-)

**Comment by NeQuissimus** [comment]
This is what I do to add XMonad to any config of mine:
https://github.com/NeQuissimus/DevSetup/blob/master/nixos-xmonad.nix#L42

I am already defining that I want to enable xmonad. Why should I have to add it again to "default"?

**Comment by Ma27** [comment]
I added an option `services.xserver.plainX` to explicitly allow this.

> Can we give the modules something like default = mkDefault "xmonad"; ?
 
Sounds good, will think about this, however my I'm not sure how many time I'll have for this in the next days...

**Comment by nbp** [comment]
How is that possible to have both default to none, when there is one window manager enabled? I thought this code used to pick the only active element as the default.

I guess we should back it out, and figure out the issue with the default computation.

**Comment by Ma27** [comment]
enable a WM/DM != setting a default value.
I think that's why it has been suggested to set a default when enabling a WM/DM.

**Comment by nbp** [comment]
I think the default should be the only enabled WM/DM if any, and otherwise be the first one, or none if multiple.

**Comment by Ma27** [comment]
makes sense, will open a PR in the next days for that...

**Pull request merged**